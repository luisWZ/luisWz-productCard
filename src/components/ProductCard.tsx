import React, { CSSProperties, createContext } from 'react';
import { useProduct } from '../hooks/useProduct';
import type {
  InitialValues,
  Product,
  ProductCardHandlers,
  onChangeArgs,
} from '../interfaces';

import s from '../styles/styles.module.css';

interface ProductContext {
  count: number;
  handleCount: (factor: number) => void;
  product: Product;
  maxCount?: number;
}

export const ProductContext = createContext({} as ProductContext);
const { Provider } = ProductContext;

export interface ProductCardProps {
  children: (args: ProductCardHandlers) => JSX.Element;
  product: Product;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const countHandler = useProduct({ onChange, product, value, initialValues });

  return (
    <Provider
      value={{
        ...countHandler,
        product,
        maxCount: initialValues?.maxCount,
      }}
    >
      <div className={`${s.productCard} ${className}`} style={style}>
        {children({
          ...countHandler,
          product,
        })}
      </div>
    </Provider>
  );
};
