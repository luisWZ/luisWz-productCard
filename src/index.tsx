export * from './components';

export { Product } from './interfaces';
