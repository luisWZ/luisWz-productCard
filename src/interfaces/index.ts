export interface Product {
  id: number;
  description: string;
  image?: string;
}

export interface onChangeArgs {
  product: Product;
  count: number;
}

export interface Cart extends Product {
  count: number;
}

export interface InitialValues {
  count?: number;
  maxCount?: number;
}

export interface ProductCardHandlers {
  count: number;
  isMaxCountReached: boolean;
  maxCount?: number;
  product: Product;

  handleCount: (factor: number) => void;
  reset: () => void;
}
