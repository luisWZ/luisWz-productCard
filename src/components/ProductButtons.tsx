import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';

import s from '../styles/styles.module.css';

export interface ProductButtonsProps {
  className?: string;
  style?: CSSProperties;
}

export const Buttons = ({ className, style }: ProductButtonsProps) => {
  const { count, handleCount, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(() => !!maxCount && count >= maxCount, [count, maxCount]);

  return (
    <div className={`${s.buttonsContainer} ${className}`} style={style}>
      <button disabled={count === 0} className={s.buttonMinus} onClick={() => handleCount(-1)}>
        -
      </button>
      <div className={s.countLabel}>{count}</div>
      <button disabled={isMaxReached()} className={s.buttonAdd} onClick={() => handleCount(1)}>
        +
      </button>
    </div>
  );
};
