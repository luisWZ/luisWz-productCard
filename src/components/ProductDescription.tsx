import React, { useContext, CSSProperties } from 'react';
import { ProductContext } from './ProductCard';

import s from '../styles/styles.module.css';

export interface ProductDescriptionProps {
  description?: string;
  className?: string;
  style?: CSSProperties;
}

export const Description = ({ description, className, style }: ProductDescriptionProps) => {
  const { product } = useContext(ProductContext);
  const descriptionToShow = description ? description : product.description;

  return (
    <span className={`${s.productDescription} ${className}`} style={style}>
      {descriptionToShow}
    </span>
  );
};
