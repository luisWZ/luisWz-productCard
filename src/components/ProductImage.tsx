import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';

import noImg from '../assets/no-image.jpg';
import s from '../styles/styles.module.css';

export interface ProductImageProps {
  image?: string;
  className?: string;
  style?: CSSProperties;
}

export const Image = ({ image, className, style }: ProductImageProps) => {
  const { product } = useContext(ProductContext);

  const imageToShow = image ?? product.image ?? noImg;
  const alt = product?.description && (image || product.image) ? product.description : 'Product';

  return (
    <img className={`${s.productImg} ${className}`} style={style} src={imageToShow} alt={alt} />
  );
};
