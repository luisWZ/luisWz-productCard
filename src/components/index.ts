import { ProductCard as ProductCardHOC, ProductCardProps as ProductProps } from './ProductCard';

import {
  Description as ProductCardDescription,
  ProductDescriptionProps,
} from './ProductDescription';
import { Image as ProductCardImage, ProductImageProps } from './ProductImage';
import { ProductButtonsProps, Buttons as ProductCardButtons } from './ProductButtons';

export { Description as ProductCardDescription } from './ProductDescription';
export { Image as ProductCardImage } from './ProductImage';
export { Buttons as ProductCardButtons } from './ProductButtons';

export interface ProductCardProps {
  (props: ProductProps): JSX.Element;
  Description: (props: ProductDescriptionProps) => JSX.Element;
  Image: (props: ProductImageProps) => JSX.Element;
  Buttons: (props: ProductButtonsProps) => JSX.Element;
}

export const ProductCard: ProductCardProps = Object.assign(ProductCardHOC, {
  Description: ProductCardDescription,
  Image: ProductCardImage,
  Buttons: ProductCardButtons,
});
