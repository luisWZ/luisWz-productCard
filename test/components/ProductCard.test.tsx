import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { products } from '../../example/data/products';

describe('ProductCard', () => {
  it('should render', () => {
    const wrapper = renderer.create(
      <ProductCard product={products[0]}>
        {() => (
          <>
            <ProductCard.Image />
            <ProductCard.Description />
            <ProductCard.Buttons />
          </>
        )}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should increase the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={products[0]}>
        {({ handleCount, count }) => (
          <>
            <h1>{count}</h1>
            <button onClick={() => handleCount(1)}></button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();

    renderer.act(() => {
      (tree as any).children[1].props.onClick();
    });

    tree = wrapper.toJSON();

    expect((tree as any).children[0].children[0]).toBe('1');
  });
});
