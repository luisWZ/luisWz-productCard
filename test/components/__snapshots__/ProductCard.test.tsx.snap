// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should render 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffee Mug"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
  <span
    className="productDescription undefined"
  >
    Coffee Mug
  </span>
  <div
    className="buttonsContainer undefined"
  >
    <button
      className="buttonMinus"
      disabled={true}
      onClick={[Function]}
    >
      -
    </button>
    <div
      className="countLabel"
    >
      0
    </div>
    <button
      className="buttonAdd"
      disabled={false}
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;
